documentation/secure_storage_rpmb.md

```diff
@@ -0,0 +1,108 @@
+ E         v          v            :         (tee_svc_storage.c)
+ R    eMMC driver   OP-TEE driver  : (tee_rpmb_fs.c, tee_fs_key_manager.c)
+ N                           ^     :            (tee_rpmb.c)
+ E                           |     :                  ^
+ L                           |     :                  |
+                             |     :                  |
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~|~~~~~~~~~~~~~~~~~~~~~~~~|~~~~~~~~~~~~~~~~~~~~
+                             v                        v
+                           Secure monitor / EL3 firmware
+```
+
+## The Secure Storage API
+
+This part is common with the REE-based filesystem. The interface between the
+system calls in [core/tee/tee_svc_storage.c](../core/tee/tee_svc_storage.c) and
+the RPMB filesystem is the **tee_file_operations** structure `tee_file_ops`.
+
+## The RPMB filesystem
+
+The FS implementation is mainly in [core/tee/tee_rpmb_fs.c](../core/tee/tee_rpmb_fs.c).
+
+The RPMB partition is divided in three parts:
+
+- The first 128 bytes are reserved for partition data (struct
+**rpmb_fs_partition**).
+- At offset 512 is the File Allocation Table (FAT). It is an array of
+struct **rpmb_fat_entry** elements, one per file. The FAT grows dynamically as
+files are added to the filesystem. Among other things, each entry has the start
+address for the file data, its size, and the filename.
+- Starting from the end of the RPMB partition and extending downwards is the
+file data area.
+
+Space in the partition is allocated by the general-purpose allocator functions:
+`tee_mm_alloc()` and `tee_mm_alloc2()`.
+
+## Device access
+
+The device access layer is in [core/arch/arm/tee/tee_rpmb.c](../core/arch/arm/tee/tee_rpmb.c).
+
+There is no eMMC controller driver in OP-TEE. The device operations, such as:
+programming the security key, reading the write counter value, reading or writing
+blocks of data, all have to go through the normal world. They are handled by the
+`tee-supplicant` process which further relies on the kernel's `ioctl()`
+interface to access the device. `tee-supplicant` also has an emulation mode
+which implements a virtual RPMB device for test purposes.
+
+RPMB operations are initiated on request from the FS layer. Memory buffers for
+requests and responses are allocated in shared memory using
+`thread_optee_rpc_alloc_payload()`.
+Buffers are passed to the normal world in a `TEE_RPC_RPMB_CMD` message, thanks
+to the `thread_rpc_cmd()` function. Most RPMB requests and responses use the
+data frame format defined by the JEDEC eMMC specification.
+
+HMAC authentication is implemented here also.
+
+## Encryption
+
+The FS encryption routines are in [core/tee/tee_fs_key_manager.c](../core/tee/tee_fs_key_manager.c).
+
+Block encryption is enabled by default to protect file data. It may be turned
+off with `CFG_ENC_FS=n`. The algorithm is 128-bit AES in counter (CTR) mode.
+
+- During OP-TEE initialization, a 128-bit AES Secure Storage Key (SSK) is
```

Comment:
Actually, you mention below using CTR mode, which combined with an IV would allow someone seeing two different variants to be able to 'xor' them and see the 'xor' of the two blocks. (CTR mode works like a stream cypher, which is completely weak if the IV is reused). It'd be better if you used CBC, but even that can be used to reveal common prefixes of the data that are unmodified.